### ISSUE TITLE
capnproto SerializeToString?

### ISSUE THREAD

**Issue by fengmao31** [capnproto/capnproto#1648]
how to use capnproto Serialize To String？
I want to replace the past protobuf interface in the system. The interface is based on string tansport. It use the function SerializeToString  in protobuf. I want to replace it.

**Comment by fengmao31** [comment]
I read them. the capnproto cannot support in the past discussion. if I copy the FlatArray to string. it need another copy. I donnot know the loss of performance. I want to use the capnpproto to faster my transport. but if I copy the FlatArray to string, I donnot know which one is faster.
Also, it is looks like that protobuf donnot need more copy to string format.
![image](https://user-images.githubusercontent.com/19343842/227414507-300d27de-7bb0-4455-baac-735dad606e9f.png)
![image](https://user-images.githubusercontent.com/19343842/227414567-7718aff2-22fb-41db-930b-c6aecbeb9a15.png)

**Comment by kentonv** [comment]
You cannot serialize to an `std::string` without making a copy. This is a limitation of `std::string`, not Cap'n Proto. I do not recommend using `std::string` to store serialized bytes.

**Comment by fengmao31** [comment]
It looks like that protobuf support it in the lower layer without copy. and xml can support it too. I think it can bring flexibility in some simple iot system and reuse old interface.

Pass parameters (ptr, size) from application layer to the lower layer is not beautiful and not easy to maintenance.

Also, if I use the dds in the lower layer to use its shm and udp exchange meessage with other process. I can put the string as a item in the dds's CDR serialization.

**Comment by kentonv** [comment]
No, protobuf does not support serializing to std::string without a copy. Protobuf always requires a copy when serializing regardless of the destination. Since Protobuf needs a copy anyway when serializing, it can copy directly into the target std::string at the same time, so that one copy serves both purposes. With Cap'n Proto, since there's no copy involved in serializing in the first place, then you will have to make a separate copy in order to put it in an std::string.

FWIW the code you screenshotted above is mostly code I wrote.

**Comment by fengmao31** [comment]
Yes. I know it. I means that protobuf only have one copy, but if I want to get string in the capnproto, I need copy twice---copy that FlatArray copy to string(). 

Thank you. I will test the performance of two method---copy capnproto to string, protobuf. or I use a structure instead of string and remove the dds in my system.Maybe I luckily find a way to transport the message by the dds message idl define.

by the way, ros2 robot system say a new way to build a bridge between two serialization. Thety copy the context in a old ros msg serialization to dds CDR serialization. and in their document, they test it for most serialization. (of cource capnproto have no serialization process). By the method, ros2 don't need to deserialize the ros msg serialization and serialize into the dds CDR serialization and save the time. 

It is hard for me to understand the method in the low layer and operation in the memory. Maybe you have interest in know and research on this way of processing in robot community.

https://design.ros2.org/articles/serialization.html
https://design.ros2.org/articles/ros_middleware_interface.html
https://wiki.ros.org/rosserial
https://design.ros2.org/articles/legacy_interface_definition.html
https://design.ros2.org/articles/idl_interface_definition.html

Also, the promise model and other function in capnproto is good to the encapsulate the node layer and reduce the communication between node. for example, many chatbot nodes in the robot system.  today, the main node need communication too much between chatbot nodes. If they can auto-finish in a node pipline, it will be more efficient and interesting.

**Comment by fengmao31** [comment]
OK. Thank you. I will try to see how to implement it.
